plugins/catalog-backend/src/service/CatalogBuilder.ts

```diff
@@ -0,0 +1,523 @@
+  EntityPolicyProcessor,
+  FileReaderProcessor,
+  GithubOrgReaderProcessor,
+  GithubReaderProcessor,
+  GitlabApiReaderProcessor,
+  GitlabReaderProcessor,
+  HigherOrderOperation,
+  HigherOrderOperations,
+  LocationProcessor,
+  LocationReaders,
+  LocationRefProcessor,
+  PlaceholderProcessor,
+  PlaceholderResolver,
+  StaticLocationProcessor,
+  UrlReaderProcessor,
+  YamlProcessor,
+} from '../ingestion';
+import { CatalogRulesEnforcer } from '../ingestion/CatalogRules';
+import { LdapOrgReaderProcessor } from '../ingestion/processors/LdapOrgReaderProcessor';
+import {
+  jsonPlaceholderResolver,
+  textPlaceholderResolver,
+  yamlPlaceholderResolver,
+} from '../ingestion/processors/PlaceholderProcessor';
+
+export type CatalogEnvironment = {
+  logger: Logger;
+  database: PluginDatabaseManager;
+  config: Config;
+  reader: UrlReader;
+};
+
+/**
+ * A builder that helps wire up all of the component parts of the catalog.
+ *
+ * The touch points where you can replace or extend behavior are as follows:
+ *
+ * - Reader processors can be added or replaced. These implement the
+ *   functionality of reading raw data from a location, in the form of an
+ *   entity definition file.
+ * - Parser processors can be added or replaced. These accept the raw data as
+ *   read by the previous processors and parse it into raw structured data
+ *   (for example, from a binary buffer containing yaml text, to a JS object
+ *   structure).
+ * - Placeholder resolvers can be replaced or added. These run on the raw
+ *   structured data between the parsing and pre-processing steps, to replace
+ *   dollar-prefixed entries with their actual values (like $file).
+ * - Pre-processors can be added or replaced. These take the raw unvalidated
+ *   data from the parser processors and can enrich or extend it before
+ *   validation. This is the place where for example codeowners data can be
+ *   injected into partial entity definitions.
+ * - Entity policies can be added or replaced. These are the first line of
+ *   validation from the output of the pre-processing step. All policies are
+ *   given the chance to inspect the entity, and all of them have to pass in
+ *   order for the entity to be considered valid from an overall point of
+ *   view.
+ * - Field format validators can be replaced. These check the format of
+ *   individual core fields such as metadata.name, such that they adhere to
+ *   certain rules.
+ * - Entity kinds can be added or replaced. These are the second line of
+ *   validation that is applied after the entity policies, which add additional
+ *   kind-specific validation (usually based on a schema). Only one of the
+ *   entity kinds has to accept the entity, but if none of them do, the
+ *   entity is rejected as a whole.
+ * - Post-processors can be added or replaced. These take the validated
+ *   entities out of the validation step and can perform additional actions on
+ *   them.
+ */
+export class CatalogBuilder {
```

Comment:
Do we need all this? Can we have the public API be `readers -> processors -> validation`? where readers are basically processors without input, processors process entities, and validation is where all registered entity policies are applied.

Reply:
Pre is before you have a valid entity, post has the guarantee that they passed validation. Maybe not worth?

Reply:
Do we have some clear use-cases that require both? Keeping in mind my opinion that processors should do their own validation of input. Without that I think we obviously would need pre and post steps. Ofc open to discuss that opinion too, but I'm feeling it makes more sense than to keep it apart.

Reply:
It's a hassle of you can't know that there actually exists a metadata field, that you can stick the name into a URL safely, that... Etc etc. I'm just envisioning a ton of duplicated effort and opaque undefined reference errors down the line without it.

I'd be up for defining explicit interfaces for the different types of processor. They might have different interfaces and capabilities.

Reply:
Basic validation of the entity would still be done upfront. This would be for the specific fields that the processors reads and would otherwise provide custom validation for.

I think the use-case for separating the two is when you want shared field extensions, like two different processors reading `spec.foo`. Question is if it's really worth the hassle and instead just let them each validate their own input. 

There's ofc also the risk that there's for example a desync in versions or w/e between the validator and the processor, leading to errors in the processor because the validator didn't validate what the processor expected it to.

I'm thinking we can help out by relying on `yup` or `JSONSchema` for validation in the processors, and making that code easy to copy-paste. Still leaning towards `JSONSchema` because every time I've used `yup` I've been frustrated with it only doing 90% of what I want x)

Reply:
I still don't know how all thus fits with needing the ability to emit only partially valid entities and stuff. I'm confused. Shall we really redesign things from the ground up as part of this?

Reply:
I mean this doesn't even effectively change behaviors. It's a different way of wiring up things better than we did. So I'm not sure I'm ready to derail this more than this now

Reply:
I'm bringing it up now because this PR takes some conventions that are loosely encoded in the default processors and hardens them into a very specific API with much less wiggle room and potential of experimentation and evolution.